compiler/rustc_typeck/src/check/generator_interior.rs

```diff
@@ -409,3 +447,184 @@ impl<'a, 'tcx> Visitor<'tcx> for ArmPatCollector<'a> {
         }
     }
 }
+
+// Returns whether it emitted a diagnostic or not
+// Note that this fn and the proceding one are based on the code
+// for creating must_use diagnostics
+pub fn check_must_not_suspend_ty<'tcx>(
+    fcx: &FnCtxt<'_, 'tcx>,
+    ty: Ty<'tcx>,
+    hir_id: HirId,
+    expr: Option<&'tcx Expr<'tcx>>,
+    source_span: Span,
+    yield_span: Span,
+    descr_pre: &str,
+    descr_post: &str,
+    plural_len: usize,
+) -> bool {
+    if ty.is_unit()
+    // || fcx.tcx.is_ty_uninhabited_from(fcx.tcx.parent_module(hir_id).to_def_id(), ty, fcx.param_env)
+    // FIXME: should this check is_ty_uninhabited_from
+    {
+        return true;
+    }
+
+    let plural_suffix = pluralize!(plural_len);
+
+    match *ty.kind() {
+        ty::Adt(..) if ty.is_box() => {
+            let boxed_ty = ty.boxed_ty();
+            let descr_pre = &format!("{}boxed ", descr_pre);
+            check_must_not_suspend_ty(
+                fcx,
+                boxed_ty,
+                hir_id,
+                expr,
+                source_span,
+                yield_span,
+                descr_pre,
+                descr_post,
+                plural_len,
+            )
+        }
+        ty::Adt(def, _) => check_must_not_suspend_def(
+            fcx.tcx,
+            def.did,
+            hir_id,
+            source_span,
+            yield_span,
+            descr_pre,
+            descr_post,
+        ),
+        ty::Opaque(def, _) => {
```

Comment:
This only checks for the traits on the opaque type, but it may make sense to additionally reveal the opaque type similar to how `Send` and `Sync` are leaked through opaque types.

Alternatively, with type alias impl trait, it may be useful to allow adding the attribute to the declaring type alias.

Reply:
This lint only triggers on an `impl Trait` if the trait itself is marked as `must_not_suspend`. If the value made opaque by that `impl Trait` (or `dyn Trait`) is `must_not_suspend`, that is not caught

I can definitely add a comment to add TAIT support though!